pyop2/codegen/rep2loopy.py

```diff
@@ -0,0 +1,781 @@
+        preamble = preamble + "\n" + code
+
+    wrapper = loopy.register_preamble_generators(wrapper, [_PreambleGen(preamble)])
+
+    # register petsc functions
+    wrapper = loopy.register_function_id_to_in_knl_callable_mapper(
+        wrapper, petsc_function_lookup)
+
+    return wrapper
+
+
+def argtypes(kernel):
+    args = []
+    for arg in kernel.args:
+        if isinstance(arg, loopy.ValueArg):
+            args.append(as_ctypes(arg.dtype))
+        elif isinstance(arg, loopy.ArrayArg):
+            args.append(ctypes.c_voidp)
+        else:
+            raise ValueError("Unhandled arg type '%s'" % type(arg))
+    return args
+
+
+@singledispatch
+def statement(expr, context):
+    raise AssertionError("Unhandled statement type '%s'" % type(expr))
+
+
+@statement.register(DummyInstruction)
+def statement_dummy(expr, context):
+    new_children = tuple(expression(c, context.parameters) for c in expr.children)
+    return DummyInstruction(expr.label, new_children)
+
+
+@statement.register(When)
+def statement_when(expr, context):
+    condition, stmt = expr.children
+    context.conditions.append(expression(condition, context.parameters))
+    stmt = statement(stmt, context)
+    context.conditions.pop()
+    return stmt
+
+
+@statement.register(Accumulate)
+def statement_assign(expr, context):
+    lvalue, _ = expr.children
+    if isinstance(lvalue, Indexed):
+        context.index_ordering.append(tuple(i.name for i in lvalue.index_ordering()))
+    lvalue, rvalue = tuple(expression(c, context.parameters) for c in expr.children)
+    within_inames = context.within_inames[expr]
+
+    id, depends_on = context.instruction_dependencies[expr]
+    predicates = frozenset(context.conditions)
+    return loopy.Assignment(lvalue, rvalue, within_inames=within_inames,
+                            predicates=predicates,
+                            id=id,
+                            depends_on=depends_on, depends_on_is_final=True)
+
+
+@statement.register(FunctionCall)
+def statement_functioncall(expr, context):
+
+    from loopy.symbolic import SubArrayRef
+    from loopy.types import OpaqueType
+
+    parameters = context.parameters
+
+    free_indices = set(i.name for i in expr.free_indices)
+    writes = []
+    reads = []
+    for access, child in zip(expr.access, expr.children):
+        var = expression(child, parameters)
+        if isinstance(var, pym.Subscript):
+            # tensor argument
+            indices = []
+            sweeping_indices = []
+            for index in var.index_tuple:
+                indices.append(index)
+                if isinstance(index, pym.Variable) and index.name in free_indices:
+                    sweeping_indices.append(index)
+            arg = SubArrayRef(tuple(sweeping_indices), var)
+        else:
+            # scalar argument or constant
+            arg = var
+        if access is READ or (isinstance(child, Argument) and isinstance(child.dtype, OpaqueType)):
```

Comment:
OK, so for `OpaqueType`s (pointers), we just have to lie to loopy about any actual access descriptor and just say `READ`.